x/staking/cache/entry.go

```diff
@@ -0,0 +1,344 @@
+	UnbondingDelegations Type = "unbonding_delegations"
+	Redelegations        Type = "redelegations"
+)
+
+type Entry[V ~[]E, E any] struct {
+	mu           sync.RWMutex
+	storeService corestoretypes.MemoryStoreService
+
+	max uint
+
+	loadFromStore func(ctx context.Context) (map[string]V, error)
+	entryType     Type
+}
+
+func NewEntry[V ~[]E, E any](
+	storeService corestoretypes.MemoryStoreService,
+	max uint,
+	loadFromStore func(ctx context.Context) (map[string]V, error),
+	entryType Type,
+) *Entry[V, E] {
+	if storeService == nil {
+		panic(fmt.Sprintf("storeService is nil for entry type %s", entryType))
+	}
+	if loadFromStore == nil {
+		panic(fmt.Sprintf("loadFromStore is nil for entry type %s", entryType))
+	}
+	entry := &Entry[V, E]{
+		storeService:  storeService,
+		max:           max,
+		loadFromStore: loadFromStore,
+		entryType:     entryType,
+	}
+	return entry
+}
+
+func (e *Entry[V, E]) getAll(ctx context.Context, cdc codec.BinaryCodec, logger func(ctx context.Context) log.Logger) (map[string]V, error) {
+	e.mu.Lock()
+	defer e.mu.Unlock()
+
+	if err := e.checkReload(ctx, cdc, logger); err != nil {
+		return nil, err
+	}
+
+	result := make(map[string]V)
+
+	store := e.storeService.OpenMemoryStore(ctx)
+	prefix := e.getPrefix()
+	iter, err := store.Iterator(prefix, storetypes.PrefixEndBytes(prefix))
+	if err != nil {
+		return nil, err
+	}
+	defer iter.Close()
+
+	prefixLen := len([]byte(e.entryType))
+	for ; iter.Valid(); iter.Next() {
+		key := string(iter.Key()[prefixLen:])
+
+		value, err := unmarshal[V](cdc, e.entryType, iter.Value())
+		if err != nil {
+			return nil, err
+		}
+
+		result[key] = value
+	}
+
+	return result, nil
+}
+
+func (e *Entry[V, E]) get(ctx context.Context, cdc codec.BinaryCodec, key string, logger func(ctx context.Context) log.Logger) (V, error) {
+	e.mu.Lock()
+	defer e.mu.Unlock()
+
+	if err := e.checkReload(ctx, cdc, logger); err != nil {
+		return make(V, 0), err
+	}
+
+	store := e.storeService.OpenMemoryStore(ctx)
+	storeKey := e.getStoreKey(key)
+
+	bz, err := store.Get(storeKey)
+	if err != nil {
+		return make(V, 0), err
+	}
+
+	if bz == nil {
+		return make(V, 0), nil
+	}
+
+	return unmarshal[V](cdc, e.entryType, bz)
+}
+
+func (e *Entry[V, E]) set(ctx context.Context, cdc codec.BinaryCodec, key string, value V) error {
+	e.mu.Lock()
+	defer e.mu.Unlock()
+	store := e.storeService.OpenMemoryStore(ctx)
+
+	metadata, err := e.getMetadata(store, cdc)
+	if err != nil {
+		return err
+	}
+
+	if metadata.IsFull {
+		return types.ErrCacheMaxSizeReached
+	}
+
+	exist := false
+	if e.max > 0 {
+		var err error
+		exist, err = e.exists(store, key)
+		if err != nil {
+			return err
+		}
+	}
+
+	bz, err := marshal(cdc, e.entryType, value)
+	if err != nil {
+		return err
+	}
+
+	if err := e.setStore(store, key, bz); err != nil {
+		return err
+	}
+
+	if e.max > 0 && !exist {
+		count, err := e.count(store)
+		if err != nil {
+			return err
+		}
+
+		if count >= uint64(e.max) {
+			metadata.IsDirty = true
```

Comment:
Technically The cache is not `Dirty` when it reach `Max`, but it become dirty the next time the `Set` is called and when it is already Full, in that case you should skip the set and just set Dirty to True

Otherwise those two booleans do the same thing